### PR TITLE
Refactor contextual menus and Toolbar, add contextual Menu in Content empty space.

### PR DESCRIPTION
**Pull request by sBouzols**
Finally Allow directory contextual menu opening in empty space inside directory-content component.

Signed-off-by: sBouzols <sylvain.bouzols@gmail.com>

### CODE REVIEW

File: src/components/menus/content-contextual-menu-controller.js [hunk 1]
```diff
@@ -0,0 +1,556 @@
...
+        children,
+    } = props;
+    const userId = useSelector((state) => state.user.profile.sub);
+    const intl = useIntl();
+    const { enqueueSnackbar } = useSnackbar();
+
+    const DownloadIframe = 'downloadIframe';
+
+    const [openDialog, setOpenDialog] = useState(null);
+    const [lastError, setLastError] = React.useState('');
```
Comment (laurgarn) [inline src/components/menus/content-contextual-menu-controller.js]: Why two different style calls to React.useState ?

File: src/components/menus/directory-tree-contextual-menu-controller.js [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+};
+
+const DirectoryTreeContextualMenuController = (props) => {
+    const { directory, open, onClose, children } = props;
+    const userId = useSelector((state) => state.user.profile.sub);
+
+    const intl = useIntl();
+
+    const [openDialog, setOpenDialog] = useState(null);
+    const [lastError, setLastError] = React.useState('');
```
Comment (JacquesBorsenberger) [inline src/components/menus/directory-tree-contextual-menu-controller.js]: Why React.useState ? 
useState shoud be enough

File: src/components/menus/directory-tree-contextual-menu-controller.js [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+                    setHideMenu(true);
+                    handleOpenDialog(DialogsId.ADD_NEW_STUDY);
+                },
+                icon: <AddIcon fontSize="small" />,
+            });
+
+            menuItems.push({
+                messageDescriptorId: 'createNewContingencyList',
+                callback: () => {
+                    setHideMenu(true);
```
Comment (JacquesBorsenberger) [inline src/components/menus/directory-tree-contextual-menu-controller.js]: can setHideMenu be put in handleOpenDialog ?

File: src/components/menus/content-contextual-menu-controller.js (hunk 1 above)
Comment (sBouzols) [reply src/components/menus/content-contextual-menu-controller.js]: Mostly code moves. There are 40 occurences of 'React.useState' and 55 of ' useState'.
It's ok for me to homogenize this but maybe in a separate PR ?

File: src/components/menus/directory-tree-contextual-menu-controller.js (hunk 2 above)
Comment (sBouzols) [reply src/components/menus/directory-tree-contextual-menu-controller.js]: see response https://github.com/gridsuite/gridexplore-app/pull/76#discussion_r811280040

File: src/components/menus/directory-tree-contextual-menu-controller.js (hunk 3 above)
Comment (sBouzols) [reply src/components/menus/directory-tree-contextual-menu-controller.js]: IT must and it does now ! :)

**Review by JacquesBorsenberger — APPROVED** [review]

File: src/components/menus/content-contextual-menu-controller.js (hunk 1 above)
Comment (JacquesBorsenberger) [reply src/components/menus/content-contextual-menu-controller.js]: ok

**Comment by SlimaneAmar** [comment]
when we put the focus on an element and we display the menu the focus disappears and the delete button appears

File: src/components/dialogs/copy-to-script-dialog.js [hunk 4]
```diff
@@ -40,7 +40,7 @@ const CopyToScriptDialog = ({
         setNewNameValue(event.target.value);
     };
     const handleClick = () => {
-        onClick(id, newNameValue, selectedDirectory);
+        onClick(id, newNameValue, selectedDirectory.elementUuid);
```
Comment (SlimaneAmar) [inline src/components/dialogs/copy-to-script-dialog.js]: Is it used the param `selectedDirectory.elementUuid` ?

File: src/components/dialogs/replace-with-script-dialog.js [hunk 5]
```diff
@@ -32,7 +32,7 @@ const ReplaceWithScriptDialog = ({ id, open, onClose, onClick, title }) => {
     };
 
     const handleClick = () => {
-        onClick(id, selectedDirectory);
+        onClick(id, selectedDirectory.elementUuid);
```
Comment (SlimaneAmar) [inline src/components/dialogs/replace-with-script-dialog.js]: Same above

File: src/components/menus/content-contextual-menu.js [hunk 6]
```diff
@@ -0,0 +1,519 @@
...
+    COPY_FILTER_TO_SCRIPT_CONTINGENCY: 'copy_filter_to_script_contingency',
+    REPLACE_FILTER_BY_SCRIPT: 'replace_filter_by_script',
+    COPY_FILTER_TO_SCRIPT: 'copy_filter_to_script',
+    GENERIC_FILTER: 'generic_filter',
+    NONE: 'none',
+};
+
+const ContentContextualMenu = (props) => {
+    const {
+        directory,
```
Comment (SlimaneAmar) [inline src/components/menus/content-contextual-menu.js]: is a prop ?

File: src/components/menus/content-contextual-menu.js [hunk 7]
```diff
@@ -0,0 +1,519 @@
...
+        activeElement,
+        selectedElements,
+        open,
+        onClose,
+        children,
```
Comment (SlimaneAmar) [inline src/components/menus/content-contextual-menu.js]: Same

File: src/components/menus/content-contextual-menu.js [hunk 8]
```diff
@@ -0,0 +1,519 @@
...
+    const handleLastError = useCallback(
+        (message) => {
+            enqueueSnackbar(message, {
+                variant: 'error',
+            });
+        },
+        [enqueueSnackbar]
+    );
+
+    const handleOpenDialog = (DialogId) => {
```
Comment (laurgarn) [inline src/components/menus/content-contextual-menu.js]: Here, I would find it less missleading to use 'dialogId' over 'DialogId'

File: src/utils/custom-hooks.js [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+    );
+
+    const fetchData = useCallback(
+        async (finalParams) => {
+            dispatch({ type: FetchStatus.FETCHING });
+            try {
+                // Params resolution
+                const response = await fetchFunction.apply(
+                    null,
+                    Object.keys(finalParams).map((key) => {
```
Comment (laurgarn) [inline src/utils/custom-hooks.js]: The 'return' is for using the statement form, for dealing with the 80 characters at max a line ? Why not use Object.values() ?

**Review by laurgarn — APPROVED** [review]

File: src/components/menus/content-contextual-menu.js (hunk 8 above)
Comment (sBouzols) [reply src/components/menus/content-contextual-menu.js]: fixed

File: src/utils/custom-hooks.js (hunk 9 above)
Comment (sBouzols) [reply src/utils/custom-hooks.js]: fixed

**Comment by sBouzols** [comment]
Fixed the delete button appearance in the toolbar.
For focus design which is lost when ctx menu is opened, it is the default behaviour and it's already present in main branch.
The workaround/patch (a dirty one), used in the TreeView can't be easily applied here. Will see later.

File: src/components/dialogs/copy-to-script-dialog.js (hunk 4 above)
Comment (sBouzols) [reply src/components/dialogs/copy-to-script-dialog.js]: You are right, it's useless here

File: src/components/dialogs/replace-with-script-dialog.js (hunk 5 above)
Comment (sBouzols) [reply src/components/dialogs/replace-with-script-dialog.js]: here too removed

File: src/components/menus/content-contextual-menu.js (hunk 6 above)
Comment (sBouzols) [reply src/components/menus/content-contextual-menu.js]: You are right. Removed, and use `selectedDirectory` from redux instead in this component

File: src/components/menus/content-contextual-menu.js (hunk 7 above)
Comment (sBouzols) [reply src/components/menus/content-contextual-menu.js]: a remnant of old design HOC, removed

**Pull request merged**